cmd/executor/main.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/gorilla/websocket"
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/zajca/go-executor/internal/pkg/cleaner"
+	"github.com/zajca/go-executor/internal/pkg/client"
+	"github.com/zajca/go-executor/internal/pkg/job"
+)
+
+// with more thant 1K limit process are throwing "invalid memory address or nil pointer dereference"
+// this might be system limit for processes
+// when limit is reached system will not proccess new messages
+const JOB_COUNT_LIMIT = 1000
+
+type applicationStatus int
+
+const (
+	Starting applicationStatus = 0
+	Running  applicationStatus = 1
+	Draining applicationStatus = 2
+)
+
+func (s applicationStatus) String() string {
+	switch s {
+	case Starting:
+		return "STARTING"
+	case Running:
+		return "RUNNING"
+	case Draining:
+		return "DRAINING"
+	}
+	panic("Unknow status")
+}
+
+type healthResponse struct {
+	Status           applicationStatus
+	Jobs             []job.JobResponse
+	CountClearing    uint32
+	CountJobsRunning uint32
+	CountJobsTotal   uint32
+}
+
+var (
+	upgrader = websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		},
+	}
+	jobDone     = make(chan *job.Job)
+	jobCleared  = make(chan *job.Job)
+	broadcaster = client.NewBroadcaster()
+	cleanerInst = cleaner.NewJobCleaner(broadcaster)
+	jobRunner   = job.Runner{}
+	cmdPath     = strings.Split(os.Getenv("CMD_PATH"), ",")
+	msgPath     = os.Getenv("MSG_PATH")
+	status      = Starting
+)
+
+func ws(c echo.Context) error {
+	ws, err := upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		return err
+	}
+	broadcaster.AddClient(ws)
+	defer ws.Close()
+
+	for {
+		// top limit is not exact as there could be more messages by time this is evaluated
+		if jobRunner.RunningJobsCount() >= JOB_COUNT_LIMIT {
+			continue
+		}
+		if status == Draining {
+			continue
+		}
+		// Read
+		_, msg, err := ws.ReadMessage()
+		if err != nil {
+			continue
+		}
+		c.Logger().Debug(msg)
+		currentJob, err := job.MakeJob(cmdPath, msgPath, string(msg))
+		if err != nil {
+			c.Logger().Error(err)
+			continue
+		}
+		c.Logger().Debug(currentJob.JobId)
+		//init job dumper
+		currentDumper := job.NewJobDumper(&currentJob)
+		currentDumper.Init()
+		// init job channels
+		messages := make(chan *job.Message)
+		PID := make(chan int, 1)
+		cmdRunDone := make(chan bool, 1)
+		// run job
+		go jobRunner.Run(&currentJob, messages, PID, cmdRunDone, c.Logger())
+
+		go func() {
+			for m := range messages {
+				isSend := broadcaster.SendMessage(m, c.Logger())
```

Comment:
Tato `go routine` vyzera trocha podozrivo.

V Go bezi go routine, kym sa sama neukonci (alebo nastane `kill` celeho programu).

V `main` riesis nejakym sposobom `interrupt`, ale malo by sa to nejak dostat do kazdej `goroutine`.
Respektive aspon do tych, kde sa robi nejaka vacsi praca a typicky je tam `for` alebo nejaka blokujuca operacia.

Na to sa obycajne pouziva `select { .... <-ctx.Done()`, tj. pred kazdou iteraciou by si skontroloval, ... ci sa ma este pokracovat dalej, a program sa nahodou neukoncuje.

https://dev.to/mcaci/how-to-use-the-context-done-method-in-go-22me

Ak by si chcel nejaky jednoduchy priklad, tak napr.:
- program, spracuj co stihnes za `10 sekund` a potom sa pekne ukonci
- `ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)`
- ten `ctx` dostanes do `goroutine`
- a tam bude `select`

Je o tom kopec clankov na nete. 

Iny priklad:
https://github.com/keboola/keboola-as-code/blob/5c532154f9a29c106c093e1c8a5b07c17f3e4a65/internal/pkg/http/client/pool.go#L145-L147